106-auto-rebalance-on-imbalanced-clusters.md

```diff
@@ -0,0 +1,524 @@
+It is also the end state of a previous successfully completed or failed auto-rebalancing.
+In case of successful completion, once the rebalance moves to `Ready` state, we will delete the KafkaRebalance and move the empty the `anomaly_list` then update the `auto-rebalance` state to `Idle`.
+In case of failed auto-rebalancing, once the rebalance moves to `NotReady` state, we will follow the same procedure we used in successful completion.
+In this state, the operator removes the finalizer and deletes the corresponding "actual" `KafkaRebalance` custom resource.
+
+#### RebalanceOnAnomalyDetection
+
+In this state, an anomaly was detected and a corresponding ConfigMap was generated by the notifier.
+
+ A KafkaRebalance resource will now be applied to the cluster to fix the imbalanced cluster. This kafka rebalance will be based on the template provided by the user, if no template is provided then the kafkaRebalance will be created with default configurations.
+
+```mermaid
+flowchart TB
+  A[KafkaRebalanceState] --> B{if state}
+  B --> C[Ready]
+  C -- queued scale Down --> D[RebalanceOnScaleDown]
+  C -- queued scale up --> E[RebalanceOnScaleUp]
+  C -- queued rebalance --> F[RebalanceOnAnomalyDetection]
+  B --> G[PendingProposal, ProposalReady or Rebalancing]
+  G --> H{{rebalancing is running}}
+  B --> I[NotReady]
+  I --> J{{Transition to Idle state and delete rebalance resource and empty the `anomaly_list`}}
+```
+
+Checking the current `KafkaRebalance` status:
+
+* if `Ready`, the rebalance was successful.
+  * if there is a queued rebalancing scale down (`Kafka.status.autoRebalance.modes[remove-brokers]` exists), start the rebalancing scale down and transition to **RebalanceOnScaleDown**.
+  * if there is a queued rebalancing scale up (`Kafka.status.autoRebalance.modes[add-brokers]` exists), start the rebalancing scale up and transition to **RebalanceOnScaleUp**.
+  * If no queued rebalancing scale down or scale up, just transition to **Idle**, clean `Kafka.status.autoRebalance.modes`, delete the "actual" `KafkaRebalance` custom resource.
+* if `PendingProposal`, `ProposalReady` or `Rebalancing`, the rebalancing is still running.
+  * No further actions required.
+* if `NotReady`
+  * the rebalancing failed, transition to **Idle** and also removing the corresponding mode from the status. The operator also deletes the "actual" `KafkaRebalance` custom resource and empties the ConfigMap.
+
+If, during an ongoing auto-rebalancing, the `KafkaRebalance` custom resource is not there anymore on the next reconciliation, it could mean the user deleted it while the operator was stopped/crashed/not running.
+In this case, the FSM will assume it as `NotReady` so falling in the last case above.
+
+## Affected/not affected projects
+
+This change will affect the Strimzi cluster operator and a new repository named `strimzi-notifier` will be added under the Strimzi organisation.
```

Comment:
+1 of separate repository for the notifier. But that should be likely already detailed in earlier in the proposal.